### PR TITLE
Add query params to supply summary endpoint

### PR DESCRIPTION
**Pull request by akhilkumarpilli**
- Fullfil CMC's API requirements. The old supply summary endpoint now takes optional `query` parameters to filter and send the required details. Optional query parameters are:
`q` => `total` or `circulating`
`denom` => denom to filter the amount (required if `q` is supplied)
`decimals` => decimals to convert the tokens (required if `q` is supplied)

### CODE REVIEW

File: x/supply/client/rest/query.go [hunk 1]
```diff
@@ -19,11 +28,40 @@ func RegisterRoutes(ctx context.CLIContext, r *mux.Router) {
...
+				metric = res.Total
+			} else if q == "circulating" {
+				metric = res.Circulating
+			} else {
+				rest.WriteErrorResponse(w, http.StatusBadRequest, "wrong q value")
+				return
+			}
+			supply := sdk.Coin{}
+			for _, coin := range metric {
+				if coin.Denom == denom {
```
Comment (anilcse) [inline x/supply/client/rest/query.go]: Can we use `StakingDenom` for the comparison here? This should be stitchable to any cosmos based network

File: x/supply/client/rest/query.go [hunk 2]
```diff
@@ -10,6 +12,13 @@ import (
 
 	"github.com/ovrclk/cosmos-supply-summary/x/supply/query"
 	"github.com/ovrclk/cosmos-supply-summary/x/supply/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+var (
+	decimals float64 = 6
+	denom            = "uakt"
```
Comment (anilcse) [inline x/supply/client/rest/query.go]:
```suggestion
```

File: x/supply/client/rest/query.go [hunk 3]
```diff
@@ -19,11 +28,40 @@ func RegisterRoutes(ctx context.CLIContext, r *mux.Router) {
...
-			return
+		q := r.URL.Query().Get("q")
+		if len(q) == 0 {
+			res, err := query.NewRawClient(ctx, types.ModuleName).CirculatingSupply()
+			if err != nil {
+				rest.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
+				return
+			}
+			rest.PostProcessResponse(w, ctx, res)
+		} else {
```
Comment (anilcse) [inline x/supply/client/rest/query.go]: We can avoid `else` here

```suggestion
			rest.PostProcessResponse(w, ctx, res)
			return
		}
```

File: x/supply/client/rest/query.go [hunk 4]
```diff
@@ -19,11 +28,40 @@ func RegisterRoutes(ctx context.CLIContext, r *mux.Router) {
...
+				rest.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
+				return
+			}
+			metric := sdk.Coins{}
+			if q == "total" {
+				metric = res.Total
+			} else if q == "circulating" {
+				metric = res.Circulating
+			} else {
+				rest.WriteErrorResponse(w, http.StatusBadRequest, "wrong q value")
```
Comment (anilcse) [inline x/supply/client/rest/query.go]:
```suggestion
				rest.WriteErrorResponse(w, http.StatusBadRequest, "wrong query value, allowed values are `total`, `circulating`")
```

File: x/supply/client/rest/query.go (hunk 1 above)
Comment (anilcse) [reply x/supply/client/rest/query.go]: If not, let's take the denom in the query itself.

**Review by anilcse — DISMISSED** [review]

File: x/supply/client/rest/query.go [hunk 5]
```diff
@@ -19,11 +24,57 @@ func RegisterRoutes(ctx context.CLIContext, r *mux.Router) {
...
+		if len(denom) == 0 {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, "`denom` value is required")
+			return
+		}
+		if len(decString) == 0 {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, "`decimal` value is required")
+			return
+		}
+
+		decimals, err := strconv.ParseFloat(decString, 64)
```
Comment (anilcse) [inline x/supply/client/rest/query.go]: Should we ensure that it is an `integer`

File: x/supply/client/rest/query.go [hunk 6]
```diff
@@ -19,11 +24,57 @@ func RegisterRoutes(ctx context.CLIContext, r *mux.Router) {
...
 		}
-		rest.PostProcessResponse(w, ctx, res)
+		metric := sdk.Coins{}
+		if q == "total" {
+			metric = res.Total
+		} else if q == "circulating" {
+			metric = res.Circulating
+		} else {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, "wrong query value, allowed values are `total`, `circulating`")
+			return
```
Comment (anilcse) [inline x/supply/client/rest/query.go]: use switch-case

**Review by anilcse — APPROVED** [review]
lgtm

**Review by boz — APPROVED** [review]
Looks great, thanks @akhilkumarpilli.

I think it'd be nice to not have to cast to `float64` or even `int64` for this - the bignum types should be able to be formatted correctly, but that can be done another time.  Maybe it's not a big deal, but I try to stay away from floats if at all possible.

**Comment by anilcse** [comment]
> Looks great, thanks @akhilkumarpilli.
> 
> I think it'd be nice to not have to cast to `float64` or even `int64` for this - the bignum types should be able to be formatted correctly, but that can be done another time. Maybe it's not a big deal, but I try to stay away from floats if at all possible.

Agreed, @akhilkumarpilli let's have a follow-up PR for this

**Pull request merged**